dart/lib/src/http_client/sentry_http_client.dart

```diff
@@ -0,0 +1,99 @@
+import 'package:http/http.dart';
+
+import '../../sentry.dart';
+
+/// A [http](https://pub.dev/packages/http)-package compatible HTTP client
+/// which records requests as breadcrumbs.
+///
+/// Remarks:
+/// If this client is used as a wrapper, a call to close does not close the
+/// given client.
+///
+/// The `SentryHttpClient` can be used as a standalone client like this:
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+///
+/// var client = SentryHttpClient();
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+///
+/// The `SentryHttpClient` can also be used as a wrapper for your own
+/// HTTP [Client](https://pub.dev/documentation/http/latest/http/Client-class.html):
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+/// import 'package:http/http.dart' as http;
+///
+/// final myClient = http.Client();
+///
+/// var client = SentryHttpClient(client: myClient);
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+//
+// Possible enhancement:
+// Track the time the HTTP request took.
+// For example with Darts Stopwatch:
+// https://api.dart.dev/stable/2.10.4/dart-core/Stopwatch-class.html
+class SentryHttpClient extends BaseClient {
+  SentryHttpClient({Client client, Hub hub}) {
+    _hub = hub ?? HubAdapter();
+    if (client != null) {
+      _client = client;
+      _userdefinedClient = true;
+    } else {
+      _client == Client();
+    }
+  }
+
+  Client _client;
+  Hub _hub;
+  bool _userdefinedClient = false;
+
+  @override
+  Future<StreamedResponse> send(BaseRequest request) async {
+    final response = await _client.send(request);
+    // See https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/
+    // for how the breadcrumb should look like
+    _hub.addBreadcrumb(
+      Breadcrumb(
+        // TODO: should this always be `http`?
+        // Otherwise use http or https depending on the request.
+        type: request.url.scheme,
+        data: {
+          'url': request.url.toString(),
+          'method': request.method,
+          'status_code': response.statusCode,
+          // reason is optional, therefor only add it in case it is not null
+          if (response.reasonPhrase != null) 'reason': response.reasonPhrase,
+        },
+      ),
+    );
+    return response;
+  }
+
+  @override
+  void close() {
+    // In case the user did not specify a client himself
+    // we should close the client we initialized.
+    //
+    // If the user did specifiy a client himself,
+    // the user is in charge of closing his own client.
+    // Thus we don't close it here.
```

Comment:
tbh I'm wondering why we need this and not simply `_client.close()`

so if the user did not specify a client and call `close()`, obviously we need to close it.

if the user did specify a client, for which reason they would be calling `close` then and not only in their own Client?

Actually, When they are wrapping up their own Client in our Wrapper, they should be using the Wrapper only and their own Client won't be used anymore, so I see the common behavior to actually call `close()` in the wrapper and not to the original Client, unless I am missing something.

Reply:
I'm fine with that. I'll point this out in the documentation then.

Reply:
I think the limitation would be if they have their own Client being shared across multiple wrappers, and they wanna release the allocation of the Sentry wrapper, that would be a problem, but tbh, I think this should still be fixable by their own code, a Wrapper ideally would mimic a real implementation which is `closing` its Client.

Reply:
Yeah, that was also my original reasoning for implementing it that way. However I haven't done that myself in any of my projects, so I guess it's more of a niche feature.